opal/corelib/hash.rb

```diff
@@ -762,57 +762,56 @@ def length
 
   alias member? has_key?
 
-  def merge(other, &block)
-    dup.merge!(other, &block)
+  def merge(*others, &block)
+    dup.merge!(*others, &block)
   end
 
-  def merge!(other, &block)
-    %x{
-      if (!other.$$is_hash) {
-        other = #{Opal.coerce_to!(other, Hash, :to_hash)};
-      }
+  def merge!(*others, &block)
+    others.map! { |other| Opal.coerce_to!(other, Hash, :to_hash) }
+
+    `var i, other_keys, length, key, value, other_value;`
+    others.each do |other|
+      %x{
+        other_keys = other.$$keys, length = other_keys.length;
```

Comment:
Unless there's a good reason I think it's better to keep the iteration in a JS `for` loop, to keep it as fast as possible.

`map!` and `merge` both introduce additional checks and overhead, but since the array is coming from `*others` we can expect it to always be a JS array (vs. Enumerator etc.).

A single for loop is also probably the implementation used by MRI:

```
>> (a= {}).merge!({a: 1}, 1, "a")
Traceback (most recent call last):
        6: from /Users/elia/.asdf/installs/ruby/3.0.0/bin/irb:23:in `<main>'
        5: from /Users/elia/.asdf/installs/ruby/3.0.0/bin/irb:23:in `load'
        4: from /Users/elia/.asdf/installs/ruby/3.0.0/lib/ruby/gems/3.0.0/gems/irb-1.3.0/exe/irb:11:in `<top (required)>'
        3: from (irb):1:in `<main>'
        2: from (irb):2:in `rescue in <main>'
        1: from (irb):2:in `merge!'
TypeError (no implicit conversion of Integer into Hash)
>> a
=> {:a=>1}
>> 
```

As you can see the first item is merged even if the second raises an error.

Reply:
Thanks for your pointing out!

>A single for loop is also probably the implementation used by MRI:

I didn't dig MRI implementation 🙇 Agreed!

Updated in https://github.com/opal/opal/pull/2187/commits/f429e1efca40eae18513c94272cb1b2f36e4d558 🙏